### PR TITLE
Improve Getting Started Documentation — No Binary Approach

### PR DESCRIPTION
**Pull request by mathewbyrne**
Addresses #408 — this is a set of improvements to the getting started documentation.  The main change is deprecating using a global binary, and moving to always using `dep` to ensure users have a consistent version.

#415 discusses the whys for this approach.

### CODE REVIEW

File: docs/content/getting-started.md [hunk 1]
```diff
@@ -6,32 +6,54 @@ menu: main
...
+Assuming you already have a working [Go environment](https://golang.org/doc/install), create a directory for the project in your `$GOPATH`:
+
+```sh
+$ mkdir -p $GOPATH/src/github.com/[username]/gqlgen-todos
+```
+
+Add the following file to your project under `scripts/gqlgen.go`:
+
+```go
+// +build ignore
```
Comment (vektah) [inline docs/content/getting-started.md]: there needs to be an empty line between this comment and package main for the build tag to work

File: docs/content/getting-started.md [hunk 2]
```diff
@@ -60,17 +82,17 @@ type Mutation {
...
+This has created an empty skeleton with all files you need:
 
- - gqlgen.yml - The gqlgen config file, knobs for controlling the generated code.
- - generated.go - The graphql execution runtime, the bulk of the generated code
- - models_gen.go - Generated models required to build the graph. Often you will override these with models you write yourself. Still very useful for input types.
- - resolver.go - This is where your application code lives. generated.go will call into this to get the data the user has requested. 
+ - `gqlgen.yml` — The gqlgen config file, knobs for controlling the generated code.
+ - `generated.go` — The GraphQL execution runtime, the bulk of the generated code.
+ - `models_gen.go` — Generated models required to build the graph. Often you will override these with your own models. Still very useful for input types.
+ - `resolver.go` — This is where your application code lives. `generated.go` will call into this to get the data the user has requested.
```
Comment (vektah) [inline docs/content/getting-started.md]: did this also create a server entry point?

File: docs/content/getting-started.md [hunk 3]
```diff
@@ -242,34 +264,14 @@ query findTodos {
...
-
-```bash
-dep init
-dep ensure
-```
-
 At the top of our resolvers.go a go generate command was added that looks like this:
 ```go
-//go:generate gorunpkg github.com/99designs/gqlgen
+//go:generate go run ./scripts/gqlgen.go
```
Comment (vektah) [inline docs/content/getting-started.md]: did you update the template for resolvers.go?

Comment (mathewbyrne) [reply docs/content/getting-started.md]: Ah interesting.  Yeah what should we do there?  We can't assume users will have a script to run at `scripts/gqlgen.go`.

File: cmd/root.go [hunk 4]
```diff
@@ -9,6 +9,9 @@ import (
 	"github.com/99designs/gqlgen/graphql"
 	"github.com/99designs/gqlgen/internal/gopath"
 	"github.com/urfave/cli"
+
+	// Required since otherwise dep will prune away these unused packages before codegen has a chance to run
+	_ "github.com/99designs/gqlgen/handler"
```
Comment (mathewbyrne) [inline cmd/root.go]: This seems to be enough to get `init` to run cleanly.  Attempting to run `server/server.go` afterwards however still requires some additional deps (websockets I think).  So I added a `dep ensure` after `gqlgen init` in the docs.

**Comment by emil-nasso** [comment]
Is it possible to add an explanation to what `./...` means? 

This is not unlikely to be the first run in with `go generate` for many users. The go documentation is not very good or clear on this.

To my understanding it runs go generate in all subdirectories so `./...` means "the subdirectories (...) of the current directory (.)".

I think it would help to alleviate some confusion to just have a short explaination like:

> You can now run any generation commands for all subdirectories in your project with:

**Comment by mathewbyrne** [comment]
@emil-nasso yeah I think that's a fair comment.  I'll add a note.

File: docs/content/getting-started.md [hunk 5]
```diff
@@ -88,22 +117,27 @@ type Todo struct {
 }
 ```
 
-And then tell gqlgen to use this new struct by adding this to the gqlgen.yml:
+Next tell gqlgen to use this new struct by adding it to `gqlgen.yml`:
+
 ```yaml
 models:
   Todo:
     model: github.com/vektah/gqlgen-tutorials/gettingstarted.Todo
```
Comment (samjbobb) [inline docs/content/getting-started.md]: Earlier the doc says:

> `mkdir -p $GOPATH/src/github.com/[username]/gqlgen-todos`

So I think this doesn't match. Probably should be:

`model: github.com/[username]/gqlgen-todos.Todo`

File: docs/content/getting-started.md [hunk 6]
```diff
@@ -152,17 +186,16 @@ better than generating.
 ### Write the resolvers
 
 This is a work in progress, we have a way to generate resolver stubs, but it only cant currently update existing code. We can force it to run again by deleting `resolvers.go` and re-running gqlgen:
+
 ```bash
-rm resolvers.go
-gqlgen
+$ rm resolvers.go
```
Comment (samjbobb) [inline docs/content/getting-started.md]: For me, this is `resolver.go` no `resolvers.go`

File: docs/content/getting-started.md [hunk 7]
```diff
@@ -152,17 +186,16 @@ better than generating.
...
+$ go run scripts/gqlgen.go
 ```
 
-Now we just need to fill in the `not implemented` parts
-
+Now we just need to fill in the `not implemented` parts.  Update `graph/graph.go`
```
Comment (samjbobb) [inline docs/content/getting-started.md]: I think (for me) this should be "Update `resolver.go`"

File: docs/content/getting-started.md [hunk 8]
```diff
@@ -242,34 +275,14 @@ query findTodos {
...
-
-*Note*: If you dont have dep installed yet, you can get it [here](https://github.com/golang/dep)
-
-First uninstall the global version we grabbed earlier. This is a good way to prevent version mismatch footguns.
-
-```bash
-rm ~/go/bin/gqlgen
-rm -rf ~/go/src/github.com/99designs/gqlgen
-``` 
+At the top of our `resolvers.go` add the following line:
```
Comment (samjbobb) [inline docs/content/getting-started.md]: `resolvers.go` -> `resolver.go` (for me at least, I think)

**Comment by samjbobb** [comment]
Hey thanks for this project and this PR. This updated documentation really helped me get started.

I created a new project and set my `Gopkg.toml` to:

```
[[constraint]]
  name = "github.com/99designs/gqlgen"
  branch = "improved-getting-started"
```

And then ran through the Getting Started guide. 

Overall, was successful and the new docs helped. I noticed a couple of discrepancies, so I added some commends to the PR. Hope that's helpful!

Thanks again.

**Pull request merged**